zetaclient/orchestrator/chain_activate.go

```diff
@@ -0,0 +1,220 @@
+package orchestrator
+
+import (
+	"time"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	btcobserver "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/observer"
+	btcrpc "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/rpc"
+	btcsigner "github.com/zeta-chain/zetacore/zetaclient/chains/bitcoin/signer"
+	evmobserver "github.com/zeta-chain/zetacore/zetaclient/chains/evm/observer"
+	evmsigner "github.com/zeta-chain/zetacore/zetaclient/chains/evm/signer"
+	"github.com/zeta-chain/zetacore/zetaclient/chains/interfaces"
+	"github.com/zeta-chain/zetacore/zetaclient/common"
+	"github.com/zeta-chain/zetacore/zetaclient/config"
+)
+
+// WatchEnabledChains watches for run-time chain activation and deactivation
+func (oc *Orchestrator) WatchEnabledChains() {
+	oc.logger.Std.Info().Msg("WatchChainActivation started")
+
+	ticker := time.NewTicker(common.ZetaBlockTime)
+	for {
+		select {
+		case <-ticker.C:
+			oc.ActivateDeactivateChains()
+		case <-oc.stop:
+			oc.logger.Std.Info().Msg("WatchChainActivation stopped")
+			return
+		}
+	}
+}
+
+// ActivateDeactivateChains activates or deactivates chain observers and signers
+func (oc *Orchestrator) ActivateDeactivateChains() {
+	// create new signer and observer maps
+	// Note: the keys of the two maps are chain IDs and they are always exactly matched
+	newSignerMap := make(map[int64]interfaces.ChainSigner)
+	newObserverMap := make(map[int64]interfaces.ChainObserver)
+
+	// create new signers and observers
+	oc.CreateObserversEVM(newSignerMap, newObserverMap)
+	oc.CreateObserversBTC(newSignerMap, newObserverMap)
+
+	// loop through existing observer map to deactivate chains that are not in new observer map
+	for chainID, observer := range oc.observerMap {
+		_, found := newObserverMap[chainID]
+		if !found {
+			oc.logger.Std.Info().Msgf("ActivateDeactivateChains: deactivating chain %d", chainID)
+
+			observer.Stop()
+
+			// remove signer and observer from maps
+			oc.mu.Lock()
+			delete(oc.signerMap, chainID)
+			delete(oc.observerMap, chainID)
+			oc.mu.Unlock()
+		}
+	}
+
+	// loop through new observer map to activate chains that are not in existing observer map
+	for chainID, observer := range newObserverMap {
+		_, found := oc.observerMap[chainID]
+		if !found {
+			oc.logger.Std.Info().Msgf("ActivateDeactivateChains: activating chain %d", chainID)
+
+			// open database and load data
+			err := observer.LoadDB(oc.dbPath)
+			if err != nil {
+				oc.logger.Std.Error().
+					Err(err).
+					Msgf("ActivateDeactivateChains: error LoadDB for chain %d", chainID)
+				continue
+			}
+
+			observer.Start()
+
+			// add signer and observer to maps
+			oc.mu.Lock()
+			oc.signerMap[chainID] = newSignerMap[chainID]
+			oc.observerMap[chainID] = observer
+			oc.mu.Unlock()
+		}
+	}
+}
+
+// CreateObserversEVM creates signer and observer maps for all enabled EVM chains
+func (oc *Orchestrator) CreateObserversEVM(
```

Comment:
Shouldn't we rather say `set` instead of `create`?
Also since this include the signer, I think we can name it "ChainClient"

```
SetChainClientEVM
```
Same for BTC

Reply:
It's a `create`, not a `set`. For the name, I used a long name `CreateObserverSignerEVM` during development if you don't mind verbosity.

Reply:
https://github.com/zeta-chain/node/pull/2411/commits/b52654f8856fd9ce599599a003d3c3cf3290d2de